### ISSUE TITLE
testnet btcd stuck on block 1414430, hits nil pointer error

### ISSUE THREAD

**Issue by sangaman** [btcsuite/btcd#1318]
I have a testnet btcd instance I've been using on and off for the past few months. I recently started it up again for the first time since the duplicate inputs vulnerability and the invalid blocks on testnet, and noticed that btcd would get stuck on block 1414430. (side note: I ran another node which got stuck on 1414411, updating btcd resolved that issue)

I tried updating btcd and restarting and it still stalled on 1414430. `getinfo` would give the following result:

```
btcctl --testnet getinfo
{
  "version": 120000,
  "protocolversion": 70002,
  "blocks": 1414430,
  "timeoffset": 1,
  "connections": 7,
  "proxy": "",
  "difficulty": 1,
  "testnet": true,
  "relayfee": 0.00001,
  "errors": ""
}
```

I tried restarting, and it crashed shortly after startup a nil pointer dereference. Now every time I start btcd back up I get the same error with the following output:

```
2018-10-03 23:02:59.852 [INF] BTCD: Version 0.12.0-beta
2018-10-03 23:02:59.854 [INF] BTCD: Loading block database from 'C:\Users\Daniel\AppData\Local\Btcd\data\testnet\blocks_ffldb'
2018-10-03 23:02:59.952 [INF] BCDB: Detected unclean shutdown - Repairing...
2018-10-03 23:02:59.954 [INF] BCDB: Database sync complete
2018-10-03 23:02:59.954 [INF] BTCD: Block database loaded
2018-10-03 23:02:59.970 [INF] INDX: Transaction index is enabled
2018-10-03 23:02:59.970 [INF] INDX: Committed filter index is enabled
2018-10-03 23:02:59.973 [INF] CHAN: Loading block index...
2018-10-03 23:03:14.315 [INF] CHAN: Chain state (height 1414430, hash 00000000e9ca87d7af01d7408a7d470255d38afb59ba62a5ac248af2b5aed70e, totaltx 47735451, work 3651569259393818814629)
2018-10-03 23:03:14.331 [INF] RPCS: RPC server listening on 127.0.0.1:18334
2018-10-03 23:03:14.331 [INF] RPCS: RPC server listening on [::1]:18334
2018-10-03 23:03:14.435 [INF] AMGR: Loaded 12317 addresses from file 'C:\Users\Daniel\AppData\Local\Btcd\data\testnet\peers.json'
2018-10-03 23:03:14.436 [INF] CMGR: 1 addresses found from DNS seed testnet-seed.bluematt.me
2018-10-03 23:03:14.436 [INF] CMGR: 24 addresses found from DNS seed seed.tbtc.petertodd.org
2018-10-03 23:03:14.436 [INF] CMGR: Server listening on 0.0.0.0:18333
2018-10-03 23:03:14.437 [INF] CMGR: 23 addresses found from DNS seed testnet-seed.bitcoin.jonasschnelli.ch
2018-10-03 23:03:14.437 [INF] CMGR: Server listening on [::]:18333
2018-10-03 23:03:14.554 [INF] CMGR: DNS discovery failed on seed testnet-seed.bitcoin.schildbach.de: lookup testnet-seed.bitcoin.schildbach.de: no such host
2018-10-03 23:03:14.689 [INF] PEER: Disconnecting non-segwit peer 35.156.118.148:18333 (outbound), isn't segwit enabled and we need more segwit enabled peers
2018-10-03 23:03:14.689 [INF] SYNC: New valid peer 35.156.118.148:18333 (outbound) (/Bitcoin ABC:0.14.6(EB8.0)/)
2018-10-03 23:03:14.691 [INF] SYNC: Lost peer 35.156.118.148:18333 (outbound)
2018-10-03 23:03:14.733 [INF] SYNC: New valid peer 92.53.120.80:18333 (outbound) (/Satoshi:0.16.0/)
2018-10-03 23:03:14.733 [INF] SYNC: Syncing to block height 1414965 from peer 92.53.120.80:18333
2018-10-03 23:03:15.053 [INF] SYNC: New valid peer 167.99.194.65:18333 (outbound) (/Satoshi:0.13.0/)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x68 pc=0x7a62aa]

goroutine 38 [running]:
github.com/btcsuite/btcd/database/ffldb.rollbackOnPanic(0xc04edac6c0)
        C:/Users/Daniel/go/src/github.com/btcsuite/btcd/database/ffldb/db.go:1831 +0x7e
panic(0x9f05e0, 0xf03310)
        C:/Go/src/runtime/panic.go:491 +0x291
github.com/btcsuite/btcd/blockchain/indexers.dbStoreFilterIdxEntry(0xed9240, 0xc04edac6c0, 0xeccca0, 0x10, 0x10, 0xc04200ca60, 0xc05c190980, 0x20, 0x20, 0x0, ...)
        C:/Users/Daniel/go/src/github.com/btcsuite/btcd/blockchain/indexers/cfindex.go:68 +0xba
github.com/btcsuite/btcd/blockchain/indexers.storeFilter(0xed9240, 0xc04edac6c0, 0xc0563493b0, 0xc04e27c240, 0xc04e27c200, 0x0, 0x0)
        C:/Users/Daniel/go/src/github.com/btcsuite/btcd/blockchain/indexers/cfindex.go:179 +0x2ba
github.com/btcsuite/btcd/blockchain/indexers.(*CfIndex).ConnectBlock(0xc042112400, 0xed9240, 0xc04edac6c0, 0xc0563493b0, 0xc042056a80, 0x110, 0x110, 0x0, 0x0)
        C:/Users/Daniel/go/src/github.com/btcsuite/btcd/blockchain/indexers/cfindex.go:225 +0x200
github.com/btcsuite/btcd/blockchain/indexers.dbIndexConnectBlock(0xed9240, 0xc04edac6c0, 0xed7660, 0xc042112400, 0xc0563493b0, 0xc042056a80, 0x110, 0x110, 0x0, 0x0)
        C:/Users/Daniel/go/src/github.com/btcsuite/btcd/blockchain/indexers/manager.go:89 +0x149
github.com/btcsuite/btcd/blockchain/indexers.(*Manager).ConnectBlock(0xc0420fe870, 0xed9240, 0xc04edac6c0, 0xc0563493b0, 0xc042056a80, 0x110, 0x110, 0x0, 0xc053b5d080)
        C:/Users/Daniel/go/src/github.com/btcsuite/btcd/blockchain/indexers/manager.go:508 +0xcb
github.com/btcsuite/btcd/blockchain.(*BlockChain).connectBlock.func1(0xed9240, 0xc04edac6c0, 0xc04edac6c0, 0x0)
        C:/Users/Daniel/go/src/github.com/btcsuite/btcd/blockchain/chain.go:644 +0x20f
github.com/btcsuite/btcd/database/ffldb.(*db).Update(0xc042114300, 0xc05d03c870, 0x0, 0x0)
        C:/Users/Daniel/go/src/github.com/btcsuite/btcd/database/ffldb/db.go:1888 +0xac
github.com/btcsuite/btcd/blockchain.(*BlockChain).connectBlock(0xc04203cf00, 0xc04ed6a000, 0xc0563493b0, 0xc04e92c1b0, 0xc042056a80, 0x110, 0x110, 0xc042031908, 0x41192f)
        C:/Users/Daniel/go/src/github.com/btcsuite/btcd/blockchain/chain.go:611 +0x346
github.com/btcsuite/btcd/blockchain.(*BlockChain).connectBestChain(0xc04203cf00, 0xc04ed6a000, 0xc0563493b0, 0x0, 0x0, 0x0, 0x7739ae)
        C:/Users/Daniel/go/src/github.com/btcsuite/btcd/blockchain/chain.go:1145 +0x225
github.com/btcsuite/btcd/blockchain.(*BlockChain).maybeAcceptBlock(0xc04203cf00, 0xc0563493b0, 0x0, 0x0, 0x0, 0xf16a80)
        C:/Users/Daniel/go/src/github.com/btcsuite/btcd/blockchain/accept.go:79 +0x210
github.com/btcsuite/btcd/blockchain.(*BlockChain).ProcessBlock(0xc04203cf00, 0xc0563493b0, 0xc000000000, 0xc054f70000, 0x0, 0x0)
        C:/Users/Daniel/go/src/github.com/btcsuite/btcd/blockchain/process.go:228 +0x2cb
github.com/btcsuite/btcd/netsync.(*SyncManager).handleBlockMsg(0xc042084630, 0xc053a9acc0)
        C:/Users/Daniel/go/src/github.com/btcsuite/btcd/netsync/manager.go:569 +0x20c
github.com/btcsuite/btcd/netsync.(*SyncManager).blockHandler(0xc042084630)
        C:/Users/Daniel/go/src/github.com/btcsuite/btcd/netsync/manager.go:1166 +0x58f
created by github.com/btcsuite/btcd/netsync.(*SyncManager).Start
        C:/Users/Daniel/go/src/github.com/btcsuite/btcd/netsync/manager.go:1388 +0xec
```

This is on `go version go1.9.2 windows/amd64`

**Comment by gggin** [comment]
my bitcoin-core also fault on this block, sth is happened, then I restart my bitcoin-core testnet with reindex the problem is resloved!

**Comment by Roasbeef** [comment]
Looks like you were running the version that originally was created on my fork. Since then, the indexing format for the filters has changed. As a result, you'll need to re-index in order to proceed forward in the chain.

**Comment by sangaman** [comment]
Yeah that's correct, I'll try a reindex. Will update here if for whatever reason it doesn't work.

**Comment by sangaman** [comment]
@Roasbeef maybe a dumb question, but I can't find any way to reindex btcd without just deleting everything and starting over. I tried `verifychain` but that didn't seem to fix anything. How can I do this in btcd?

**Comment by totaloutput** [comment]
@sangaman can you dropcfindex and restart without it?
    --dropcfindex           Deletes the index used for committed filtering (CF) support from the database on start up and then exits.

**Comment by sangaman** [comment]
Tried that, it dropped the index, I restarted btcd, and after some time of reindexing it hit this error:

```
2018-10-15 23:03:57.376 [ERR] BTCD: Unable to start server on [:18333]: failed to read block 0000000000000027b5c8af617f72bd6c1dd2947563b09540c7a907c38905243c from file 15, offset 19060499: EOF: EOF
```

I then tried `droptxindex` for good measure, but after restarting btcd I'm getting the same error message. This appears to be some random block from a while ago, not the block that was initially giving me trouble. I'm quite certain my hard drive is not having data corruption issues, fwiw.
